lib/pdn/pdn.c

```diff
@@ -0,0 +1,457 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <zephyr.h>
+#include <init.h>
+#include <modem/pdn.h>
+#include <modem/at_cmd.h>
+#include <modem/at_notif.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(pdn, CONFIG_PDN_LOG_LEVEL);
+
+/* Buffer size for AT command and responses.
+ * Longest AT command is:
+ *   AT+CGDCONT=20,"IPV4V6","<APN>",0,0,0,0,0,0,0,0,0
+ * where the APN can be up to 100 chars.
+ * That's 43 chars, plus 100, which we round to 160.
+ */
+#define AT_BUF_SIZE (160)
+/* Global buffer for AT commands and responses. */
+static char at_buf[AT_BUF_SIZE];
+
+#define CID_UNASSIGNED (-1)
+
+struct pdn {
+	pdn_event_handler_t callback;
+	int8_t context_id;
+};
+
+static int esm_from_notif;
+static struct k_sem notif_sem;
+static struct pdn pdn_contexts[CONFIG_PDN_CONTEXTS_MAX];
+static pdn_event_handler_t default_callback;
+
+static struct pdn *pdn_find(int cid)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pdn_contexts); i++) {
+		struct pdn *pdn = &pdn_contexts[i];
+
+		if (pdn->context_id == cid) {
+			return pdn;
+		}
+	}
+
+	return NULL;
+}
+
+static struct pdn *pdn_avail_get(void)
+{
+	return pdn_find(CID_UNASSIGNED);
+}
+
+static void at_notif_handler(void *context, const char *notif)
+{
+	char *p;
+	uint8_t cid;
+	uint8_t esm_err;
+	struct pdn *pdn;
+
+	/* These notifications are at least 10 characters long,
+	 * skip shorter strings.
+	 */
+	if (strlen(notif) < 10) {
+		return;
+	}
+
+	/* AT notif: +CNEC_ESM: <cause>,<cid>
+	 * Handle this by itself, we need to extract both CID and ESM
+	 * TODO: check before all other notifs?
+	 */
+	p = strstr(notif, "CNEC_ESM: ");
+	if (p) {
+		esm_err = atoi(p + strlen("CNEC_ESM: "));
```

Comment:
Possibly relevant: learned the other day that it's best to avoid `atoi` and use its sibling `strtol` instead due to the error handling options capability added by the latter

Reply:
Good idea, switched to `strtoul`